### PR TITLE
[expo-facebook] Disable autoinit

### LINKED ISSUE #5702: Set initialization and logging to false by default for the Facebook SDK
- Make the Expo default more private and disable "auto-logging," since most apps don't need this default. See https://developers.facebook.com/docs/app-events/getting-started-app-events-ios#add-events and https://developers.facebook.com/docs/app-events/getting-started-app-events-android/#disable-auto-events.
- Provide a method from JS to programmatically re-enable app-event logging for apps that want it. Developers can also control this on iOS via the `ios.plist` keypath in app.json. 
- I think it'd also be good to disable auto-initialization if that doesn't create technical issues. Auto-initialization is covered in the links above.

As a litmus test, when we open the Expo client or a standalone app, if it doesn't make any network requests to FB that'd be a good target.

### PR DESCRIPTION
**Pull request by sjchmiela**
# Why

Fixes https://github.com/expo/expo/issues/5702.

# How

According to https://github.com/expo/expo/issues/5702#issuecomment-540013928 and https://github.com/expo/expo/issues/5702#issuecomment-540379234.

Universe PR: https://github.com/expo/universe/pull/3716
XDL PR: https://github.com/expo/expo-cli/pull/1118

# Test Plan

Verified manually that:

- a new experience opened up doesn't autoinitialize Facebook SDK (no request is issued, Facebook SDK remains uninitialized)
- initializing manually enables login
- setting `facebookAppId` in `app.json`, `autoinit: true` in JS and restarting the experience autoinitializes the SDK

### CODE REVIEW

**Review by ide — COMMENTED** [review]
@cruzach could you review the docs in this PR to make sure they read well?
@esamelson could you make sure this gets landed since someone else will need to take it over?

File: docs/pages/versions/unversioned/sdk/facebook.md [hunk 1]
```diff
@@ -44,14 +44,38 @@ You may have to switch the app from 'development mode' to 'public mode' on the F
...
+- If you provide an explicit `appId`, it will override any other source.
+
+The same resolution mechanism works for `appName`.
+
+### `Facebook.setAutoInitEnabledAsync(enabled: boolean): Promise<void>`
+
+Sets whether Facebook SDK should autoinitialize itself. SDK initialization involves eg. fetching app settings from Facebook or a profile of the logged in user. In some cases, you may want to disable or delay the SDK initialization, such as to obtain user consent or fulfill legal obligations. This method corresponds to [this](https://developers.facebook.com/docs/app-events/getting-started-app-events-ios#disable-sdk-initialization) and [this](https://developers.facebook.com/docs/app-events/getting-started-app-events-android/#disable-sdk-initialization) native SDK methods. Even though calling this method with `enabled == true` initializes the Facebook SDK on iOS, it does not on Android and we recommend always calling `initializeAsync` before performing any actions with effects that should be visible to the user (like `loginWithPermissions`).
```
Comment (cruzach) [inline docs/pages/versions/unversioned/sdk/facebook.md]:
```suggestion
Sets whether the Facebook SDK should autoinitialize itself. SDK initialization involves eg. fetching app settings from Facebook or a profile of the logged in user. In some cases, you may want to disable or delay the SDK initialization, such as to obtain user consent or fulfill legal obligations. This corresponds to [this iOS SDK method](https://developers.facebook.com/docs/app-events/getting-started-app-events-ios#disable-sdk-initialization) and [this Android SDK method](https://developers.facebook.com/docs/app-events/getting-started-app-events-android/#disable-sdk-initialization). Even though calling this method with `enabled == true` initializes the Facebook SDK on iOS, it does not on Android and we recommend always calling `initializeAsync` before performing any actions with effects that should be visible to the user (like `loginWithPermissions`).
```

File: docs/pages/versions/unversioned/sdk/facebook.md [hunk 2]
```diff
@@ -44,14 +44,38 @@ You may have to switch the app from 'development mode' to 'public mode' on the F
...
+
+In Expo, by default, autoinitialization of the Facebook SDK is disabled. You may change this value in runtime by calling this method or customize this feature in buildtime by setting appropriate `app.json` fields. The setting value is persisted across runs (value set with this method overriddes value from buildtime).
```
Comment (cruzach) [inline docs/pages/versions/unversioned/sdk/facebook.md]:
```suggestion
In Expo, autoinitialization of the Facebook SDK is disabled by default. You may change this value at runtime by calling this method, or customize this feature at buildtime by setting the appropriate `app.json` fields. The set value persists across runs, and the value set with this method overriddes the value set in `app.json`.
```

File: docs/pages/versions/unversioned/sdk/facebook.md [hunk 3]
```diff
@@ -44,14 +44,38 @@ You may have to switch the app from 'development mode' to 'public mode' on the F
...
 
-Your Facebook application ID. [Facebook's developer documentation](https://developers.facebook.com/docs/apps/register) describes how to get one.
+In Expo, by default, automatic logging app events is disabled. You may change this value in runtime by calling this method or customize this feature in buildtime by setting appropriate `app.json` fields. The setting value is persisted across runs (value set with this method overriddes value from buildtime).
```
Comment (cruzach) [inline docs/pages/versions/unversioned/sdk/facebook.md]:
```suggestion
In Expo, automatic logging of app events is disabled by default. You may change this at runtime by calling this method, or customize this feature at buildtime by setting the appropriate `app.json` fields. The set value persists across runs, and the value set with this method overriddes the value set in `app.json`.
```

File: docs/pages/versions/unversioned/sdk/facebook.md [hunk 4]
```diff
@@ -44,14 +44,38 @@ You may have to switch the app from 'development mode' to 'public mode' on the F
...
+
+### `Facebook.setAdvertiserIDCollectionEnabledAsync(enabled: boolean): Promise<void>`
+
+Sets whether Facebook SDK should collect and attach `advertiser-id` to sent events. `advertiser-id` let you identify and target specific customers. To learn more visit [Facebook documentation](https://developers.facebook.com/docs/app-ads/targeting/mobile-advertiser-ids) describing that topic. In some cases, you may want to disable or delay the collection of `advertiser-id`, such as to obtain user consent or fulfill legal obligations. This method corresponds to [this](https://developers.facebook.com/docs/app-events/getting-started-app-events-ios#disable-advertiser-id) and [this](https://developers.facebook.com/docs/app-events/getting-started-app-events-android/#disable-advertiser-id) native SDK methods.
```
Comment (cruzach) [inline docs/pages/versions/unversioned/sdk/facebook.md]:
```suggestion
Sets whether the Facebook SDK should collect and attach `advertiser-id` to sent events. `advertiser-id` lets you identify and target specific customers. To learn more visit [Facebook's documentation](https://developers.facebook.com/docs/app-ads/targeting/mobile-advertiser-ids) describing this topic. In some cases, you may want to disable or delay the collection of `advertiser-id`, such as to obtain user consent or fulfill legal obligations. This method corresponds to [this iOS SDK method](https://developers.facebook.com/docs/app-events/getting-started-app-events-ios#disable-advertiser-id) and [this Android SDK method](https://developers.facebook.com/docs/app-events/getting-started-app-events-android/#disable-advertiser-id).
```

File: docs/pages/versions/unversioned/sdk/facebook.md [hunk 5]
```diff
@@ -44,14 +44,38 @@ You may have to switch the app from 'development mode' to 'public mode' on the F
...
+
+In Expo, by default, collecting those IDs is disabled. You may change this value in runtime by calling this method or customize this feature in buildtime by setting appropriate `app.json` fields. The setting value is persisted across runs (value set with this method overriddes value from buildtime).
```
Comment (cruzach) [inline docs/pages/versions/unversioned/sdk/facebook.md]:
```suggestion
In Expo, collecting those IDs is disabled by default. You may change this at runtime by calling this method, or customize this feature at buildtime by setting the appropriate `app.json` fields. The set value persists across runs, and the value set with this method overriddes the value set in `app.json`.
```

**Review by cruzach — COMMENTED** [review]
lgtm! just some small wording changes

File: android/expoview/src/main/java/versioned/host/exp/exponent/modules/universal/ScopedFacebookModule.java [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+      // do nothing
+    }
+
+    if (hasPreviouslySetAutoInitEnabled || manifestDefinesAutoInitEnabled) {
+      if (facebookAppId != null) {
+        FacebookSdk.setApplicationId(facebookAppId);
+        FacebookSdk.setApplicationName(facebookApplicationName);
+        FacebookSdk.sdkInitialize(context, () -> {
+          mIsInitialized = true;
+          FacebookSdk.fullyInitialize();
```
[the remark this answers is not in the document]
Comment (esamelson) [reply android/expoview/src/main/java/versioned/host/exp/exponent/modules/universal/ScopedFacebookModule.java]: This particular method is safe to call concurrently as it just sets a single static variable to `true`. I tested calling the `initializeAsync` and `setAutoInitEnabled` methods multiple times in a row in JS without await and it seemed to work just as well as if I'd called them once.

File: docs/pages/versions/unversioned/workflow/configuration.md [hunk 7]
```diff
@@ -121,13 +121,33 @@ Any extra fields you want to pass to your experience. Values are accessible via
...
+
+### `"facebookAutoInitEnabled"`
+
+Whether the Facebook SDK should be initialized automatically. The default in Expo (Client and in standalone apps) is `false`.
+
+> **ExpoKit**: To change this field, edit `Info.plist` and `AndroidManifest.xml`.
+
+### `"facebookAutoLogAppEventsEnabled"`
+
+Whether the Facebook SDK log app events automatically. If you don't set this property, Facebook's default will be used. (Applicable only to standalone apps.)
```
[the remark this answers is not in the document]
Comment (sjchmiela) [reply docs/pages/versions/unversioned/workflow/configuration.md]: Fixed in https://github.com/expo/expo/pull/5924/commits/1c2ba4f54d1e28f03dc9ad87b35c230275192140

File: docs/pages/versions/unversioned/workflow/configuration.md [hunk 8]
```diff
@@ -121,13 +121,33 @@ Any extra fields you want to pass to your experience. Values are accessible via
...
+
+> Note that Facebook SDK must be initialized for app events to work. You may autoinitialize Facebook SDK by setting `facebookAutoInitEnabled` to `true`.
+
+> **ExpoKit**: To change this field, edit `Info.plist` and `AndroidManifest.xml`.
+
+### `"facebookAdvertiserIDCollectionEnabled"`
+
+Whether the Facebook SDK should collect `advertiser-id` properties automatically. If you don't set this property, Facebook's default will be used. (Applicable only to standalone apps.)
```
[the remark this answers is not in the document]
Comment (sjchmiela) [reply docs/pages/versions/unversioned/workflow/configuration.md]: Fixed in https://github.com/expo/expo/pull/5924/commits/1c2ba4f54d1e28f03dc9ad87b35c230275192140

**Comment by sjchmiela** [comment]
### Progress update

- **Universe PR**: merged ✅ 
- **XDL PR**: responded to [Eric's comment](https://github.com/expo/expo-cli/pull/1118#pullrequestreview-306927756) ♻️ : https://github.com/expo/expo-cli/pull/1118/commits/736461b168e3a437b2c96b056e9f073e057a1959
- **Expo PR**: ready? 🚀

**Review by esamelson — APPROVED** [review]
🚀 🎉

**Comment by serhiipalash** [comment]
Hi @sjchmiela !

We have an issue with Facebook in our app. Facebook Analytics stopped working in our latest builds for iOS. 

We investigated more on it and have found that the last build where it works was built on November 13 and the first build where it doesn't work was built on November 16.

Our app uses Expo SDK 35.

I think the issue with not working Facebook Analytics in builds made on SDK 35 is related to this pull request. There is even commit `[android] Initialize Facebook SDK lazily in older FacebookModules` in it.

And I think the build servers were updated between 13 and 16 November (related to https://github.com/expo/expo/issues/6264), which affected all next builds.

So, my question is should I create a separate issue about `[ios] Initialize Facebook SDK lazily in older FacebookModules doesn't work, which blocks Facebook Analytics` and wait for the fix, or you will fix it with SDK 36 as its new functionality caused this new issue?

@ide maybe you are also interested in this comment as you are the author of issue for this pull request.

**Comment by sjchmiela** [comment]
Hey, @serhiipalash! There seems to be no change in iOS build servers between 13 and 16th of November — see [sdk35 tarball history](https://github.com/expo/turtle/commits/master/shellTarballs/ios/sdk35), so your apps should work exactly the same before and after this time period. If you still think this is a real issue with Expo, please submit a new issue. 🙂

**Pull request merged**